Add support for navigation using keys 'j' and 'k' in help popup menu.

preetmishra opened the pull request:
@neiljp, I have added two separate commits.
- The former updates labels and descriptions for 'up/k' and 'down/j' keys. 
- The latter adds support for navigation using 'j' and 'k' keys in the help menu. I have also removed the redundant 'up/down scrolls' text from the menu's title.

Fixes #517.

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -790,7 +790,11 @@ def __init__(self, controller: Any) -> None:
 
     def keypress(self, size: urwid_Size, key: str) -> str:
-        if is_command_key('GO_BACK', key) or is_command_key('HELP', key):
+        if is_command_key('SCROLL_UP', key):
```
You may wish to include the other movement keys here too - ie. 'K'/'J' equivalents?

preetmishra commented on zulipterminal/config/keys.py:
```diff
@@ -24,14 +24,14 @@
         'key_category': 'general',
     }),
-    ('PREVIOUS_MESSAGE', {
+    ('SCROLL_UP', {
```
I agree. I'll replace the tokens/labels with the suggestions you have made.

preetmishra commented:
@neiljp, Thanks for the review. I agree we should make the navigation consistent throughout the user interface which includes adding support for navigation using 'j/k' and 'J/K' in other popups as well. I will add a follow-up PR regarding this.

I have made the requested changes which include updating labels and description for the suggested keys and added support for other movement keys ('J/K'). It is ready for review.

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -792,6 +792,12 @@ def __init__(self, controller: Any) -> None:
+            key = 'up'
+        elif is_command_key('GO_DOWN', key) \
+                or is_command_key('SCROLL_DOWN', key):
+            key = 'down'
```
I would think we'd want the `SCROLL_*` keys (and `GO_TO_BOTTOM`) to do what they would normally do, not just move up/down.

preetmishra commented:
@neiljp, I have updated the navigation keys and added the tests. However, I have not written much test code so I am not sure if the proposed implementation is what we are looking for. It'd be helpful to have your insight on this to improve it further.

neiljp commented on tests/ui/test_ui_tools.py:
```diff
@@ -1101,6 +1101,15 @@ def test_keypress_goback(self, key):
+    def test_keypress_navigation(self, key):
+        size = (200, 20)
+        self.controller.keypress(size, key)
+        assert self.controller.keypress.called
```
As written, this test is always going to pass. `self.controller` is set to be a `Mock` in the first method, as the way these are currently set up (due to `autouse=True`) that method is a little like an `__init__` for each test call. So the `assert` is testing that an explicit function call you made on a mock was actually made...which it was, in the test itself :)

You likely want `self.help_view` instead of `self.controller` on line 1110. This is the object/function you're actually testing, and when reading tests you'll maybe see some kind of setup/mocking, always a function call of the thing that's being tested, and then the `assert` type checks which specify what the end result should be, given the setup and call.

You'd also want to re-think the form of the parametrize here and use  `keys_for_command` to generate lists of tuples, rather than hard-coding in all the keys in the parametrize - in particular since the `assert` code really needs to consider the 'input' key and the resulting key passed to urwid, which are not always the same.

preetmishra commented:
@neiljp, Thanks for the feedback that you have provided. I have updated the tests based on that. I have added `expected_key` to assert whether the correct key is being passed to urwid for each 'input key'.

preetmishra commented on tests/ui/test_ui_tools.py:
```diff
@@ -1101,6 +1101,19 @@ def test_keypress_goback(self, key):
+        size = (200, 20)
+        for key in keys:
+            return_value = self.help_view.keypress(size, key)
+            assert return_value == expected_key
```
Thanks for the tip. Adding a link explaining pytest's options in the README would be helpful.

I have used list comprehension to generate parameters and removed the loop from the test code. Regarding the latter point, I have pushed an implementation without the call to the keypress method on the super class for now as urwid's implementation suggests that the value returned by the super's keypress function is the key which was passed to it. 
Urwid's link: https://github.com/urwid/urwid/blob/0b3bd90b2482e2d528aa37608304783ccadb43e2/urwid/listbox.py#L934

However, if this doesn't seem convincing, I have posted another implementation in #zulip-terminal. 
Conversation link: https://chat.zulip.org/#narrow/stream/206-zulip-terminal/topic/.5Bhelp.20popup.5D.20Navigation.20using.20'j'.20and.20'k'.2E/near/825457

preetmishra commented:
@neiljp, Thanks. I have rewritten the commit messages and removed irrelevant files from the commit message. Gitlint is pretty handy. I have also amended the tests based on your input, more on that is here: https://github.com/zulip/zulip-terminal/pull/518#discussion_r387224356.

preetmishra commented:
@neiljp, I have amended the test to mock the super class and check how its `keypress` method was called. It is ready for review.

neiljp commented:
@preetmishra I just merged this - thanks for seeing this through to the end! :tada: 

I made a few very minor adjustments to the commit text and squashed the last two commits together as I indicated previously. The commits in master are be801c72a6e0296218c9e8d3302c57d412f35e54 and the preceding one.

It was good to see an alternative solution for the test in the end - I had used `itertools` in my exploration of how it could work. There are quite a few other cases of hard-coded keys that would benefit from the general approach with the helper function - see #469.

preetmishra commented:
Thanks, I got to learn more about the pytest library while working on this one. I enjoyed it.
I'll proceed with the mentioned issue next.
